Boto ec2 volume present

x-lhan opened the pull request:
### What does this PR do?
Implement #38938 feature request

### What issues does this PR fix or reference?
#38938

### New Behavior
Added new `boto_ec2.volume_present` state. To support this state two new module functions where added `boto_ec2.create_volume` and `boto_ec2.attach_volume`. 
Added a new parameter `wait_to_finish` in `boto_ec2.detach_volume` module function to indicate the time to wait for detaching volume complete, in second. This is necessary for `boto_ec2.volume_present` state to re-attach volume properly. 

The new boto_ec2.volume_present state performs the following:
1. Creates a volume, if one with the specified tag does not exist, using the supplied parameters.
2. Ensures that the volume is attached to the instance specified, attempting to detach it if necessary.

### Tests written?
No

### Note
`versionadded` comment keyword did not added

x-lhan committed:
Added new boto_ec2.volume.present state. To support this state two new module functions where added boto_ec2.create_volume and boto_ec2.attach_volume

The new boto_ec2.volume.present state performs the following:
1. Creates a volume, if one with the specified tag does not exist,
using the supplied parameters.
2. Ensures that the volume is attached to the instance specified,
attempting to detach it if necessary.

Documentation for the state/module stuff included in comments.

Fixes #38938

x-lhan committed:
Update doc to be more clear

salt-jenkins commented:
@x-lhan, thanks for your PR! By analyzing the history of the files in this pull request, we identified @tkwilliams, @ryan-lane and @tsaridas to be potential reviewers.

Commit:
Blank line at end of file

Commit:
Blank line at end of file

ryan-lane commented on salt/modules/boto_ec2.py:
```diff
@@ -1927,7 +1929,17 @@ def detach_volume(volume_id, instance_id=None, device=None, force=False,
+        ret = conn.detach_volume(volume_id, instance_id, device, force)
+        if wait_to_finish is not None:
+            wait = 0
+            while True:
```
Why while Tue here, couldn't you put the condition here directly?

`while wait > wait_to_finish:`

ryan-lane replied:
Though, if this is going to be changed to be consistent with other modules using wait_for_sync, then this should probably stay while True

ryan-lane commented on salt/modules/boto_ec2.py:
```diff
@@ -1914,6 +1914,8 @@ def detach_volume(volume_id, instance_id=None, device=None, force=False,
                  will not have an opportunity to flush file system caches nor file system meta data.
                  If you use this option, you must perform file system check and repair procedures.
+    wait_to_finish
+       (int) - The time to wait for detaching volume complete, in second.
```
From a consistency perspective, the other boto modules use wait_for_sync, which is a bool that waits until it's finished, not based on a timeout.

ryan-lane commented on salt/states/boto_ec2.py:
```diff
@@ -1240,3 +1240,205 @@ def volumes_tagged(name, tag_maps, authoritative=False, region=None, key=None,
+        Optionally specified at volume creation time; will be ignored afterward.
+        Requires 'volume_name'.
+        e.g.: arn:aws:kms:us-east-1:012345678910:key/abcd1234-a123-456a-a12b-a123b4cd56ef
+
```
You'd also want to pass wait_for_sync through the state module, so folks can choose to wait or not.

ryan-lane commented:
Thanks for the contribution, btw, this looks pretty great, overall.

x-lhan committed:
Merge branch 'develop' into boto_ec2_volume_present

x-lhan commented:
Thank you(@ryan-lane) for the code review and suggestions.
As for the `wait_to_finish` suggestion, here is my understanding:
I totally agree this pattern should match with existing code. However `wait_for_sync` may not be the proper name as well since it is only for route53 module and it is indicated "Wait for an INSYNC change status from Route53". There are also other `wait_for_deletion`(bool) and `wait_for_delete_retires`(int) pattern exist in rds and vpc module. So is it more properly to rename the `wait_to_finsh`(int) parameter in `detach_volume` to `wait_for_detachement`(bool) and the `wait_to_finish`(int) parameter in `create_volume` to `wait_for_creation`(bool)? In the state code, I also intended not to open `wait_to_finish` to the user. Since the two place use this parameter: 1. volume should be created before tagging the volume with name 2. volume should be detached before re-attached to elsewhere. Both these case require `wait_to_finsh` should always be True. So open up that parameter may leads to error. If you are agree on the renaming part, I will do the modification. Thanks again for the quick response.

ryan-lane commented:
Sounds good

On Apr 14, 2017 3:36 PM, "x-lhan" <notifications@github.com> wrote:

> Thank you(@ryan-lane <https://github.com/ryan-lane>) for the code review
> and suggestions.
> As for the wait_to_finish suggestion, here is my understanding:
> I totally agree this pattern should match with existing code. However
> wait_for_sync may not be the proper name as well since it is only for
> route53 module and it is indicated "Wait for an INSYNC change status from
> Route53". There are also other wait_for_deletion(bool) and
> wait_for_delete_retires(int) pattern exist in rds and vpc module. So is
> it more properly to rename the wait_to_finsh(int) parameter in
> detach_volume to wait_for_detachement(bool) and the wait_to_finish(int)
> parameter in create_volume to wait_for_creation(bool)? In the state code,
> I also intended not to open wait_to_finish to the user. Since the two
> place use this parameter: 1. volume should be created before tagging the
> volume with name 2. volume should be detached before re-attached to
> elsewhere. Both these case require wait_to_finsh should always be True.
> So open up that parameter may leads to error. If you are agree on the
> renaming part, I will do the modification. Thanks again for the quick
> response.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/40694#issuecomment-294231745>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABd5MkjhQASJMOzilvgQXqlmWjORy_MXks5rv9jPgaJpZM4M9VUt>
> .
>

x-lhan committed:
Renamed variable to match with existing pattern

Renamed 'wait_to_complete' variable to 'wait_for_detachement'/'wait_for_creation' in 'detach_volume'/'create_volume' boto_ec2 execution module

ryan-lane commented:
+1

cachedout commented:
Go Go Jenkins!

cachedout commented:
Go Go Jenkins!

cachedout commented:
Go Go Jenkins!

cachedout commented:
@ryan-lane Do you see outstanding changes that we need to make for this or is it good from your perspective?

ryan-lane commented:
Already put a +1 :)

Commit:
Remove extra line at end of file for lint

Commit:
Merge branch 'develop' into boto_ec2_volume_present

garethgreenaway committed:
Merge branch 'develop' into boto_ec2_volume_present
